### PR TITLE
allow creating an undef Diagonal

### PR DESCRIPTION
**Pull request by Roger-luo**
This PR mainly adds two convenient methods:

1. use `undef` to create a Diagonal matrix, similar to the `Diagonal{T}(I, n)` interface, which creates an identity.
2. `rand!` for `Diagonal`, this might not strictly follow what `rand!` currently defined (as an in-place version of `copyto!(A, rand(rng, S,  size(A)))`), but practically I think it'd be convenient to allow one initialize a `Diagonal` with random elements.  Besides, we have been pirating this function for over a year: https://github.com/QuantumBFS/LuxurySparse.jl/blob/master/src/linalg.jl#L209

### CODE REVIEW

**Comment by ViralBShah** [comment]
Seems reasonable on a first look. Need to get it to pass CI.

**Comment by mbauman** [comment]
`undef` is great, but I'm not a fan of this `rand!`.

**Comment by rfourquet** [comment]
Concerning `rand!`, this might also be related to how `rand!` works with sparse matrices, which currently makes them practically dense (although it's probably not something people want to use).

**Comment by ViralBShah** [comment]
I didn't realize, but we don't currently have `rand!` for dense matrices.

**Comment by mbauman** [comment]
You just gotta `using Random` first:

```Julia
julia> using Random

julia> rand!(zeros(3,5))
3×5 Array{Float64,2}:
 0.269889  0.153416  0.555383   0.847718  0.900311
 0.494079  0.300875  0.827245   0.132354  0.0683136
 0.627856  0.64181   0.0478133  0.948889  0.271567
```

But, yeah, the sparse behavior is exactly what I'd expect — and indeed is what currently attempts to have happen:

```Julia
julia> rand!(Diagonal(zeros(5)))
ERROR: ArgumentError: cannot set off-diagonal entry (2, 1) to a nonzero value (0.6609421826916406)
```

So that means the current implementation _can sometimes_ work if you get _very_ lucky, hah.

```Julia
julia> Random.seed!(0);

julia> rand!(Diagonal([1,2]), [0,0,42])
2×2 Diagonal{Int64,Array{Int64,1}}:
 42  ⋅
  ⋅  0
```

**Comment by Roger-luo** [comment]
Yes, the reason why we were pirating `rand!` is that we want to use it on several other sparse matrices we defined in `LuxurySparse`. And since `rand!` is missing for `Diagonal`, we just pirate it there.

> So that means the current implementation can sometimes work if you get very lucky, hah.

but this requires one to allocate a triangular matrix to assign values to `Diagonal`, which is exactly why we pirate `rand!`, the reason to use `rand!` is mainly in the case when allocation is too large (in our case it scales in 2^n)

**Comment by mbauman** [comment]
Is not the solution to do something like `rand!(nonzeros(Diagonal(...)))`? Or `rand!(@view Diagonal(...)[I])`?  Okay, neither one works (yet), but both could.

**Comment by Roger-luo** [comment]
> Is not the solution to do something like rand!(nonzeros(Diagonal(...)))? Or rand!(@view Diagonal(...)[I])? Okay, neither one works (yet), but both could.

yeah, semantically I agree this makes more sense since `rand!` should update all its input array's entries by (current) definition. I'll remove the definition of `rand!` in this PR and make another PR to support `nonzeros` on `Diagonal` then.

**Comment by ViralBShah** [comment]
Rebase to master may be necessary to pass tests.

File: stdlib/LinearAlgebra/test/diagonal.jl [hunk 1]
```diff
@@ -733,4 +733,9 @@ end
     @test dot(zeros(Int32, 0), Diagonal(zeros(Int, 0)), zeros(Int16, 0)) === 0
 end
 
+@testset "rand(undef, ::Diagonal)" begin
```
Comment (musm) [inline stdlib/LinearAlgebra/test/diagonal.jl:736]: This needs to be renamed.

Comment (Roger-luo) [reply stdlib/LinearAlgebra/test/diagonal.jl:736]: any suggestion on the name?

Comment (Roger-luo) [reply stdlib/LinearAlgebra/test/diagonal.jl:736]: oops yeah...

**Comment by Roger-luo** [comment]
ok, I'll update this and rebase it next week... got too busy this week.

**Pull request merged**